File: compile.py
#!/usr/bin/env python3

import os
import itertools
import re

# Assign directories
directory = "src/base46/lua/base46/themes"
custom = "src/themes"

highlights = ""

# Read the template.toml file and store lines after the 7th line
with open("template.toml", "r") as text_file:
    for line in itertools.islice(text_file, 7, None):
        highlights += line

# Create directories if they don't exist
if not os.path.exists("templates"):
    os.mkdir("templates")

if not os.path.exists("colors"):
    os.mkdir("colors")


def extract_base_30(lines):
    base_30 = {}
    start = False
    for line in lines:
        if line.startswith("M.base_30 = {"):
            start = True
            continue
        if start:
            if "}" in line:
                break
            key_value = line.split("=")
            if len(key_value) == 2:
                key = key_value[0].strip()
                value = key_value[1].strip().replace(",", "").replace('"', '')
                value = re.sub(r"--.*$", "", value).strip()  # Remove comments
                base_30[key] = value
    return base_30


def clean_line(line):
    # Strip extra characters and ensure proper formatting
    line = re.sub(r"--.*$", "", line).strip()  # Remove comments
    line = line.replace(",", "")
    line = line.replace("{", "")
    line = line.replace("}", "")
    line = line.replace("M.base_16 =", "")
    line = line.strip()
    return line


def process_files_in_directory(directory):
    for filename in os.listdir(directory):
        f = os.path.join(directory, filename)
        # Check if it is a file
        if os.path.isfile(f):
            themename = filename.split(".")[0].replace("-", "_")
            palette = ""
            bg = "dark"
            with open(f, "r") as text_file:
                lines = text_file.readlines()
                base_30 = extract_base_30(lines)
                started = False
                brackets = 0
                for line in lines:
                    if started:
                        if "}" in line:
                            brackets += 1
                        if "M.base_30" in line:
                            for key in base_30:
                                line = re.sub(rf"M\.base_30\.{key}\b", f'"{base_30[key]}"', line)
                        line = clean_line(line)
                        if "=" in line:
                            key, value = line.split("=")
                            key = key.strip()
                            value = value.strip()
                            if not value.startswith('"'):
                                value = f'"{value}"'
                            line = f"{key} = {value}"
                        if len(line) > 0:
                            palette += line + "\n"
                        if brackets >= 2:
                            started = False
                    if line.startswith("M.base_30 = {"):
                        started = True
                    if line.startswith("vim.opt.bg") and "light" in line:
                        bg = "light"
            template_path = "templates/" + themename + ".toml"
            with open(template_path, "w") as toml:
                toml.write("[information]\n")
                toml.write(f'name="{themename}"\n')
                toml.write('author="Name <email@domain.com>"\n')
                toml.write(f'background="{bg}"\n')
                toml.write("\n")
                toml.write("[palette]\n")
                toml.write(palette)
                toml.write("\n")
                toml.write(highlights)
            print(f"Compiling {template_path}")
            os.system(f"cd colors && colorgen-nvim -s ../{template_path}")


# Process files in both directories
process_files_in_directory(directory)
process_files_in_directory(custom)

File: compile.lua
#!/usr/bin/env lua

package.loaded["base46"] = {
  override_theme = function(M, _) return M end,
}

package.loaded["nvconfig"] = {
  ui = {
    telescope   = { style = "default" },
    cmp         = { style = "default" },
    statusline  = { theme = "default" },
  },
  base46 = { transparency = false },
}

local function read_file(path)
  local f = assert(io.open(path, "r"))
  local s = f:read("*a")
  f:close()
  return s
end

local function write_file(path, text)
  local f = assert(io.open(path, "w"))
  f:write(text)
  f:close()
end

local function scandir(dir)
  local p = assert(io.popen(('ls -1 "%s"'):format(dir), "r"))
  local t = {}

  for file in p:lines() do
    if file:match("%.lua$") then
      t[#t + 1] = dir .. "/" .. file
    end
  end

  p:close()
  return t
end

local function mkdir_p(dir)
  os.execute(('mkdir -p "%s"'):format(dir))
end

local function table_keys_sorted(t)
  local k = {}
  for key in pairs(t) do k[#k + 1] = key end
  table.sort(k)
  return k
end

local function serialize_palette(pal)
  local out = {}
  for _, k in ipairs(table_keys_sorted(pal)) do
    out[#out + 1] = ('%s = "%s"'):format(k, pal[k])
  end
  return table.concat(out, "\n") .. "\n"
end

local rename = {
  black           = "bg0",
  darker_black    = "bg1",
  black2          = "bg2",
  one_bg          = "bg3",
  one_bg2         = "bg4",
  one_bg3         = "bg5",
  white           = "fg",
  grey            = "gray1",
  grey_fg         = "gray2",
  grey_fg2        = "gray3",
  light_grey      = "gray4",
  red             = "red",
  orange          = "orange",
  yellow          = "yellow",
  sun             = "yellow2",
  green           = "green",
  vibrant_green   = "green2",
  teal            = "teal",
  cyan            = "cyan",
  blue            = "blue",
  nord_blue       = "blue2",
  purple          = "purple",
  dark_purple     = "purple2",
  pink            = "pink",
  baby_pink       = "pink2",
  statusline_bg   = "ui_statusline_bg",
  lightbg         = "ui_light_bg",
  pmenu_bg        = "ui_pmenu_bg",
  folder_bg       = "ui_folder_icon",
  line            = "ui_visual_line",
}

local THEMES_DIR  = "src/base46/lua/base46/themes"
local CUSTOM_DIR  = "src/themes"
local DEST_TPL    = "templates"
local DEST_COL    = "colors"

mkdir_p(DEST_TPL)
mkdir_p(DEST_COL)

local template = read_file("template.toml")
local header, highlights = template:match("^(.-\n\n)(.+)$")

local function process_dir(dir)
  for _, file in ipairs(scandir(dir)) do
    local name = file:match("([^/]+)%.lua$"):gsub("-", "_")
    local M    = dofile(file)

    local pal = {}
    for k, v in pairs(M.base_30 or {}) do
      pal[rename[k] or k] = v
    end

    for k, v in pairs(M.base_16 or {}) do
      pal[k] = v
    end
    local bg = "dark"
    if M.type == "light" then bg = "light"
    else
      local txt = read_file(file)
      if txt:match("vim%.opt%.bg.-light") then bg = "light" end
    end

    local tpl_body = string.format(
      '%sname="%s"\nauthor="anon"\nbackground="%s"\n\n[palette]\n%s\n%s',
      header, name, bg, serialize_palette(pal), highlights
    )

    local tpl_path = ("%s/%s.toml"):format(DEST_TPL, name)
    write_file(tpl_path, tpl_body)

    print("Compiling " .. tpl_path)
    os.execute(
      ('cd "%s" && colorgen-nvim -s ../%s'):format(DEST_COL, tpl_path)
    )
  end
end

process_dir(THEMES_DIR)
process_dir(CUSTOM_DIR)
print("All themes compiled ✔︎")
